pkg/config/remote/uptane/partial_client.go

```diff
@@ -6,21 +6,250 @@
+	metaPath, err := parseMetaPath(name)
+	if err != nil {
+		return nil, 0, err
+	}
+	if metaPath.role != roleRoot || !metaPath.versionSet {
+		return nil, 0, client.ErrNotFound{File: name}
+	}
+	for _, root := range s.roots {
+		if root.Version == metaPath.version {
+			return ioutil.NopCloser(bytes.NewReader(root.Raw)), int64(len(root.Raw)), nil
+		}
+	}
+	return nil, 0, client.ErrNotFound{File: name}
+}
+
+func (s *partialClientRemoteStore) GetTarget(path string) (stream io.ReadCloser, size int64, err error) {
+	return nil, 0, client.ErrNotFound{File: path}
+}
+
+// PartialState represents the state of a partial uptane client
+type PartialState struct {
+	RootVersion    uint64
+	TargetsVersion uint64
+}
+
 // PartialClient is a partial uptane client
 type PartialClient struct {
+	sync.Mutex
+
+	rootClient  *client.Client
+	localStore  client.LocalStore
+	remoteStore *partialClientRemoteStore
+
+	valid bool
+
+	rootVersion    uint64
+	targetsVersion uint64
+	targetMetas    data.TargetFiles
+	targetFiles    []*pbgo.File
 }
 
 // NewPartialClient creates a new partial uptane client
-func NewPartialClient() *PartialClient {
-	return &PartialClient{}
+func NewPartialClient() (*PartialClient, error) {
+	localStore := client.MemoryLocalStore()
+	err := localStore.SetMeta("root.json", json.RawMessage(meta.RootsDirector().Last()))
+	if err != nil {
+		return nil, err
+	}
+	remoteStore := &partialClientRemoteStore{}
+	c := &PartialClient{
+		rootClient:  client.NewClient(localStore, remoteStore),
+		localStore:  localStore,
+		remoteStore: remoteStore,
+		rootVersion: meta.RootsDirector().LastVersion(),
+	}
+	return c, nil
+}
+
+func (c *PartialClient) getRoot() (*data.Root, error) {
+	metas, err := c.localStore.GetMeta()
+	if err != nil {
+		return nil, err
+	}
+	rawRoot := metas["root.json"]
+	var signedRoot data.Signed
+	err = json.Unmarshal(rawRoot, &signedRoot)
+	if err != nil {
+		return nil, err
+	}
+	var root data.Root
+	err = json.Unmarshal(signedRoot.Signed, &root)
+	if err != nil {
+		return nil, err
+	}
+	return &root, nil
 }
 
-// Verify is not implemented
-func (c *PartialClient) Verify(response *pbgo.ConfigResponse) error {
-	return fmt.Errorf("not implemented")
+func (c *PartialClient) validateAndUpdateTargets(rawTargets []byte) error {
+	if len(rawTargets) == 0 {
+		return nil
+	}
+	root, err := c.getRoot()
+	if err != nil {
+		return err
+	}
+	db := verify.NewDB()
+	for _, key := range root.Keys {
+		for _, id := range key.IDs() {
+			if err := db.AddKey(id, key); err != nil {
+				return err
+			}
+		}
+	}
+	targetsRole, hasRoleTargets := root.Roles["targets"]
+	if !hasRoleTargets {
+		log.Warn("root is missing a targets role")
+		return nil
+	}
+	role := &data.Role{Threshold: targetsRole.Threshold, KeyIDs: targetsRole.KeyIDs}
+	if err := db.AddRole("targets", role); err != nil {
+		log.Warnf("could not add targets role to db: %v", err)
+		return nil
```

Comment:
Won't this break a whole product configuration deployment?

Reply:
I changed those to actual errors to make it easier to reason about